psbt/finalizer.go

```diff
@@ -0,0 +1,452 @@
+// to extract an ordered list of signatures, given
+// a serialized script (redeemscript or witness script),
+// a list of pubkeys and the signatures corresponding to those
+// pubkeys, so that the signatures will be embedded in the final
+// scriptSig or scriptWitness in the correct order.
+func extractKeyOrderFromScript(script []byte, expectedPubkeys [][]byte,
+	sigs [][]byte) ([][]byte, error) {
+	// Arrange the pubkeys and sigs into a slice of format:
+	// [[pub,sig], [pub,sig],..]
+	pubsSigs := [][][]byte{}
+	for i, pub := range expectedPubkeys {
+		tmp := [][]byte{pub, sigs[i]}
+		pubsSigs = append(pubsSigs, tmp)
+	}
+	type kv struct {
+		Key   int
+		Value [][]byte
+	}
+	var positionMap []kv
+	for _, p := range pubsSigs {
+		pos := bytes.Index(script, p[0])
+		if pos < 0 {
+			return nil, errors.New("Script does not contain pubkeys")
+		}
+		positionMap = append(positionMap, kv{Key: pos, Value: p})
+	}
+
+	sort.Slice(positionMap, func(i, j int) bool {
+		return positionMap[i].Key < positionMap[j].Key
+	})
+	// Build the return array of signatures
+	sigsNew := [][]byte{}
+	for _, x := range positionMap {
+		sigsNew = append(sigsNew, x.Value[1])
+	}
+	return sigsNew, nil
+}
+
+// getMultisigScriptWitness creates a full Witness field for the transaction,
+// given the public keys and signatures to be appended, assuming (currently
+// without validating; parseScript from txscript is not exported so it can't
+// be checked here) that the witnessScript is of type M of N multisig. This
+// is used for both p2wsh and nested p2wsh multisig cases.
+func getMultisigScriptWitness(witnessScript []byte, pubKeys [][]byte,
+	sigs [][]byte) ([]byte, error) {
+
+	orderedSigs, err := extractKeyOrderFromScript(witnessScript, pubKeys, sigs)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	var witnessItems = [][]byte{
+		nil}
+	for _, os := range orderedSigs {
+		witnessItems = append(witnessItems, os)
+	}
+	witnessItems = append(witnessItems, witnessScript)
+	err = writeTxWitness(&buf, witnessItems)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// checkSigHashFlags compares the sighash flag byte on a signature with the
+// value expected according to any PsbtInSighashType field in this section
+// of the PSBT, and returns true if they match, false otherwise.
+// If no SighashType field exists, it is assumed to be SIGHASH_ALL.
+// TODO sighash type not restricted to one byte in future?
+func checkSigHashFlags(sig []byte, input *PInput) bool {
+	expectedSighashType := txscript.SigHashAll
+	if input.SighashType != 0 {
+		expectedSighashType = input.SighashType
+	}
+
+	return expectedSighashType == txscript.SigHashType(sig[len(sig)-1])
+}
+
+// isFinalized considers this input finalized if it contains
+// at least one of the FinalScriptSig or FinalScriptWitness
+// are filled (which only occurs in a successful call to Finalize*)
+func isFinalized(p *Psbt, idx int) bool {
```

Comment:
Maybe (and very pedantic): Sometimes you use `idx` in argument to denote an input index, and other times `inIndex`. Just for consistency maybe choose one.

Reply:
OK, will fix.

Reply:
Done in 1d79a3e2102b4c8b4929d0f446de18b65453f03f